Issue #28: Unify PropagateControl and PropagateInit

BenKaufmann opened the pull request:
* update clasp
* adjust C-API:
  - drop clingo_propagate_init_add_watch{_to_thread}
  - drop clingo_propagate_init_remove_watch{_from_thread}
  - drop clingo_propagate_init_assignment
  - drop clingo_propagate_init_add_literal
  - drop clingo_propagate_init_add_clause
  - drop clingo_propagate_init_propagate
  - drop clingo_propagate_control_thread_id
  - drop clingo_propagate_control_assignment
  - add clingo_propagate_init_control for converting a
    clingo_propagate_init_t into a clingo_propagate_control_t object
  - adjust signatures of clingo_propagator functions: thread_id and
    current assignment are now explicit function parameters
  - **add clingo_propagator_attach_callback_t for initializing solving
    threads**
* adjust C++-API and tests
* adjust Python-API and tests

BenKaufmann committed:
fix some typos

BenKaufmann commented:
@rkaminsk Marked as draft for now since I'd like to get your feedback first. E.g., do you think `PropagateControl` is still a good name? Do you think the (python) API changes (in particular regarding the Propagator) are too disrupting?

rkaminsk commented:
> @rkaminsk Marked as draft for now since I'd like to get your feedback first. E.g., do you think `PropagateControl` is still a good name? Do you think the (python) API changes (in particular regarding the Propagator) are too disrupting?

Thank's I'll try to have a closer look asap. Regarding the API changes, yes I think this is fine. Since we go from clingo 5 to 6, now is the chance. Regarding the name PropagateControl, I don't have a better idea. The language model was suggesting PropgationControllor or PropagationContext but that seems very similar to me.

BenKaufmann commented:
> Regarding the name  PropagateControl, I don't have a better idea. The language model was suggesting PropgationControllor or PropagationContext but that seems very similar to me.

To me, all of these strongly hint at "propagation" but with this PR, the control is no longer limited to propagation.

I thought about "PropagatorControl" but that's not great either 🤷‍♂️

BenKaufmann commented on lib/c-api/include/clingo/propagate.h:
```diff
@@ -530,25 +454,31 @@ CLINGO_VISIBILITY_DEFAULT bool clingo_propagate_control_add_clause(clingo_propag
-typedef bool (*clingo_propagator_init_callback_t)(clingo_propagate_init_t *, void *);
+typedef bool (*clingo_propagator_init_callback_t)(clingo_assignment_t const *, clingo_propagate_init_t *, void *);
+
+//! Typedef for @ref ::clingo_propagator::attach().
```
@rkaminsk We could consider making the `thread_id` a (read-only) property of the `Assignment`. This would reduce the number of parameters in the callbacks, and it is still conceptually sound - in the context of `init()`, the assignment belongs to thread 0.

BenKaufmann replied:
Ok. Will do so.

BenKaufmann commented on lib/c-api/include/clingo/propagate.h:
```diff
@@ -402,7 +351,7 @@ CLINGO_VISIBILITY_DEFAULT bool clingo_propagate_init_add_clause(clingo_propagate
 //! @param[in] compare_equal if true compare equal instead of less than equal
 //! @param[out] result result indicating whether the problem became unsatisfiable
-//! @return wether the call was successful
+//! @return whether the call was successful
```
@rkaminsk In contrast to `clingo_propagate_init_add_minimize`, we could move `clingo_propagate_init_add_weight_constraint` to the propagate control  interface thereby making it available also to individual solver threads (during propagation/checking). Do you see any need for this?

Similarly for `clingo_propagate_init_number_of_threads`.

rkaminsk replied:
This would allow us to play with pb cuts. 👍for the threads.

BenKaufmann replied:
> 👍for the threads.

Ok. So `clingo_propagate_init_number_of_threads` will become `clingo_propagate_control_number_of_threads`.

> This would allow us to play with pb cuts.

Should I add it to this PR, then? If so, do we need the `tag` capability?

rkaminsk replied:
> `clingo_assignment_thead_id`

This is a real improvement to the interface we are building.

> `clingo_propagate_control_number_of_threads`

Now that I think about it, maybe we should leave this in the init interface. Logically, it makes most sense there. A propagator can reserve space based on this value. During propagation, it should never be necessary to use it (and in the unlikely case that it is, it can still be stored as a member).

> `clingo_propagate_init_add_weight_constraint`

If it is not too much work to get this implemented, this would be nice to have. Maybe even a nice student topic to build something on top of this.

BenKaufmann replied:
> > `clingo_propagate_control_number_of_threads`
> 
> Now that I think about it, maybe we should leave this in the init interface. 

Ups. Going to revert the change then :smile: 

> > `clingo_propagate_init_add_weight_constraint`
> 
> If it is not too much work to get this implemented, this would be nice to have. 

The amount of work necessary depends on our requirements.  If we don't need tagging (step-local weight constraints) and dynamic deletion of weight constraints, then implementation is quite trivial. Additionally adding tagging also shouldn't be too problematic. However, properly integrating weight constraints into the various deletion heuristics would require quite some effort.

rkaminsk replied:
> > > `clingo_propagate_control_number_of_threads`
> > 
> > 
> > Now that I think about it, maybe we should leave this in the init interface.
> 
> Ups. Going to revert the change then 😄

This is probably not a big thing anyway.

> 
> > > `clingo_propagate_init_add_weight_constraint`
> > 
> > 
> > If it is not too much work to get this implemented, this would be nice to have.
> 
> The amount of work necessary depends on our requirements. If we don't need tagging (step-local weight constraints) and dynamic deletion of weight constraints, then implementation is quite trivial. Additionally adding tagging also shouldn't be too problematic. However, properly integrating weight constraints into the various deletion heuristics would require quite some effort.

Then maybe we should restrict to a simple implementation first, mark the feature as experimental, and continue to refine the implementation if there is an actual need...

rkaminsk commented:
> > Regarding the name  PropagateControl, I don't have a better idea. The language model was suggesting PropgationControllor or PropagationContext but that seems very similar to me.
> 
> To me, all of these strongly hint at "propagation" but with this PR, the control is no longer limited to propagation.
> 
> I thought about "PropagatorControl" but that's not great either 🤷‍♂️

Then, let's just stick with PropagateControl. The purpose of the class goes a bit beyond simple propagation in the sense that we can also add problem clauses. But controlling propagation it is still one of its main tasks.

BenKaufmann committed:
Unify PropagateControl and PropagateInit

* update clasp
* adjust C-API:
  - drop clingo_propagate_init_add_watch{_to_thread}
  - drop clingo_propagate_init_remove_watch{_from_thread}
  - drop clingo_propagate_init_assignment
  - drop clingo_propagate_init_add_literal
  - drop clingo_propagate_init_add_clause
  - drop clingo_propagate_init_propagate
  - drop eq-support from clingo_propagate_init_add_weight_constraint
  - drop clingo_propagate_control_thread_id
  - drop clingo_propagate_control_assignment
  - add clingo_propagate_init_control for converting a
    clingo_propagate_init_t into a clingo_propagate_control_t object
  - adjust signatures of clingo_propagator functions: thread_id and
    current assignment are now explicit function parameters
  - add clingo_propagator_attach_callback_t for initializing solving
    threads
* adjust C++-API and tests
* adjust Python-API and tests

BenKaufmann commented:
@rkaminsk This should be ready now.  Weight constraints can now be added via the control object.
I prepared a clingcon branch with (minimal) required changes here: https://github.com/potassco/clingcon/tree/issue-28

One thing still missing is propagation support in the SAT-preprocessor but this is a clasp-only change.

rkaminsk commented:
> @rkaminsk This should be ready now. Weight constraints can now be added via the control object. I prepared a clingcon branch with (minimal) required changes here: https://github.com/potassco/clingcon/tree/issue-28
> 
> One thing still missing is propagation support in the SAT-preprocessor but this is a clasp-only change.

Thanks, for the effort. I'll have a look and merge next week.

rkaminsk committed:
fine-tune

- fix doxygen issues
- update python stubs
- update python dl example
- silence compiler warning
